File: reference-apps/fastapi-api-first/app/services/vault.py
"""
Vault client service for fetching secrets

Demonstrates how to integrate with HashiCorp Vault to fetch credentials
for other infrastructure services.
"""

import httpx
import logging
import re
from typing import Optional, Dict, Any
from urllib.parse import urljoin

from app.config import settings
from app.exceptions import VaultUnavailableError, ResourceNotFoundError

logger = logging.getLogger(__name__)


class VaultClient:
    """Client for interacting with HashiCorp Vault"""

    def __init__(self):
        self.vault_addr = settings.VAULT_ADDR
        self.vault_token = settings.VAULT_TOKEN
        self.headers = {"X-Vault-Token": self.vault_token}

    def _validate_secret_path(self, path: str) -> str:
        """
        Validate and sanitize the secret path to prevent SSRF attacks.

        Args:
            path: The secret path to validate

        Returns:
            Sanitized path

        Raises:
            ValueError: If path contains invalid characters
        """
        # Remove leading/trailing slashes
        path = path.strip("/")

        # Only allow alphanumeric, hyphens, underscores, and forward slashes
        if not re.match(r'^[a-zA-Z0-9/_-]+$', path):
            raise ValueError(f"Invalid secret path: {path}. Only alphanumeric characters, hyphens, underscores, and forward slashes are allowed.")

        # Prevent path traversal
        if ".." in path:
            raise ValueError(f"Invalid secret path: {path}. Path traversal sequences are not allowed.")

        return path

    async def get_secret(self, path: str, key: Optional[str] = None) -> Dict[str, Any]:
        """
        Fetch a secret from Vault KV v2 secrets engine

        Args:
            path: Secret path (e.g., 'postgres', 'mysql')
            key: Optional specific key to extract

        Returns:
            Secret data or specific key value

        Raises:
            VaultUnavailableError: If Vault is unreachable or returns an error
            ResourceNotFoundError: If the secret doesn't exist
            ValueError: If path contains invalid characters
        """
        # Validate path to prevent SSRF
        validated_path = self._validate_secret_path(path)

        # Construct URL safely
        url = urljoin(f"{self.vault_addr}/", f"v1/secret/data/{validated_path}")

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(url, headers=self.headers, timeout=5.0)

                # Handle 404 specifically
                if response.status_code == 404:
                    raise ResourceNotFoundError(
                        resource_type="secret",
                        resource_id=path,
                        message=f"Secret '{path}' not found in Vault",
                        details={"secret_path": path, "key": key}
                    )

                # Handle 403 (permission denied)
                if response.status_code == 403:
                    raise VaultUnavailableError(
                        message="Permission denied accessing Vault secret",
                        secret_path=path,
                        details={"status_code": 403}
                    )

                response.raise_for_status()

                data = response.json()
                secret_data = data.get("data", {}).get("data", {})

                if key:
                    # Check if the specific key exists
                    if key not in secret_data:
                        raise ResourceNotFoundError(
                            resource_type="secret_key",
                            resource_id=f"{path}/{key}",
                            message=f"Key '{key}' not found in secret '{path}'",
                            details={"secret_path": path, "key": key}
                        )
                    return {key: secret_data.get(key)}

                return secret_data

        except (ResourceNotFoundError, VaultUnavailableError):
            # Re-raise our custom exceptions
            raise
        except httpx.TimeoutException as e:
            logger.error(f"Timeout fetching secret from Vault: {e}")
            raise VaultUnavailableError(
                message="Timeout connecting to Vault",
                secret_path=path,
                details={"error": str(e), "timeout": "5.0s"}
            )
        except httpx.ConnectError as e:
            logger.error(f"Connection error to Vault: {e}")
            raise VaultUnavailableError(
                message="Cannot connect to Vault server",
                secret_path=path,
                details={"error": str(e), "vault_address": self.vault_addr}
            )
        except httpx.HTTPError as e:
            logger.error(f"HTTP error fetching secret from Vault: {e}")
            raise VaultUnavailableError(
                message=f"Vault returned an error: {e}",
                secret_path=path,
                details={"error": str(e)}
            )
        except Exception as e:
            logger.error(f"Unexpected error fetching secret: {e}")
            raise VaultUnavailableError(
                message=f"Unexpected error accessing Vault: {e}",
                secret_path=path,
                details={"error": str(e)}
            )

    async def check_health(self) -> Dict[str, Any]:
        """Check Vault health status"""
        url = f"{self.vault_addr}/v1/sys/health"

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{url}?standbyok=true",
                    timeout=5.0
                )

                return {
                    "status": "healthy" if response.status_code == 200 else "unhealthy",
                    "initialized": response.status_code != 501,
                    "sealed": response.status_code == 503,
                    "standby": response.status_code == 429,
                }
        except Exception as e:
            logger.error(f"Vault health check failed: {e}")
            return {
                "status": "unhealthy",
                "error": "Health check failed"
            }


# Global Vault client instance
vault_client = VaultClient()

File: reference-apps/nodejs/src/services/vault.js
/**
 * Vault Service Client
 *
 * Wrapper around node-vault for fetching secrets from HashiCorp Vault.
 * Provides a simple interface for getting service credentials.
 */

const vault = require('node-vault');
const config = require('../config');
const logger = require('../middleware/logging').logger;

class VaultClient {
  constructor() {
    this.client = vault({
      apiVersion: 'v1',
      endpoint: config.vault.address,
      token: config.vault.token,
      requestOptions: {
        timeout: config.vault.timeout
      }
    });
  }

  /**
   * Get all secrets for a service
   * @param {string} serviceName - Name of the service (e.g., 'postgres', 'redis-1')
   * @returns {Promise<Object>} Secret data
   */
  async getSecret(serviceName) {
    try {
      const path = `secret/data/${serviceName}`;
      logger.debug(`Fetching secret from Vault: ${path}`);

      const result = await this.client.read(path);

      if (!result || !result.data || !result.data.data) {
        throw new Error(`Secret not found: ${serviceName}`);
      }

      logger.debug(`Successfully fetched secret for: ${serviceName}`);
      return result.data.data;
    } catch (error) {
      logger.error(`Failed to fetch secret ${serviceName}:`, error.message);
      throw new Error(`Vault error fetching ${serviceName}: ${error.message}`);
    }
  }

  /**
   * Get a specific key from a service's secrets
   * @param {string} serviceName - Name of the service
   * @param {string} key - Specific key to retrieve
   * @returns {Promise<string>} Secret value
   */
  async getSecretKey(serviceName, key) {
    try {
      const secrets = await this.getSecret(serviceName);

      if (!(key in secrets)) {
        throw new Error(`Key '${key}' not found in ${serviceName} secrets`);
      }

      return secrets[key];
    } catch (error) {
      logger.error(`Failed to fetch secret key ${serviceName}/${key}:`, error.message);
      throw error;
    }
  }

  /**
   * Check Vault health and status
   * @returns {Promise<Object>} Health status
   */
  async healthCheck() {
    try {
      const health = await this.client.health();

      return {
        status: 'healthy',
        initialized: health.initialized || false,
        sealed: health.sealed || false,
        version: health.version || 'unknown'
      };
    } catch (error) {
      logger.error('Vault health check failed:', error.message);
      return {
        status: 'unhealthy',
        error: error.message
      };
    }
  }

  /**
   * List all secrets at a path
   * @param {string} path - Path to list (default: 'secret/metadata')
   * @returns {Promise<Array>} List of secret names
   */
  async listSecrets(path = 'secret/metadata') {
    try {
      const result = await this.client.list(path);
      return result.data.keys || [];
    } catch (error) {
      logger.error(`Failed to list secrets at ${path}:`, error.message);
      throw new Error(`Vault list error: ${error.message}`);
    }
  }
}

// Export singleton instance
const vaultClient = new VaultClient();

module.exports = {
  vaultClient,
  VaultClient
};
